File: java-spring-boot-sample-app/src/test/java/no/testframework/sampleapp/TestFrameworkIntegrationTest.java
package no.testframework.sampleapp;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

/**
 * Integration tests for the Java Spring Boot sample app.
 * Verifies auto-discovery, DI injection, and HTTP API behaviour.
 *
 * <p>Tests are ordered so each suite is only started once at a time.
 * Run-triggering tests poll for COMPLETED before the test ends to
 * avoid 409 collisions with subsequent tests.
 */
@SpringBootTest
@AutoConfigureMockMvc
@ActiveProfiles("test")
@TestMethodOrder(MethodOrderer.MethodName.class)
class TestFrameworkIntegrationTest {

    @Autowired
    private MockMvc mockMvc;

    private final ObjectMapper objectMapper = new ObjectMapper();

    // -------------------------------------------------------------------------
    // Helper — poll until COMPLETED (max 10 s)
    // -------------------------------------------------------------------------

    private void awaitCompleted(String runId) throws Exception {
        for (int i = 0; i < 100; i++) {
            Thread.sleep(100);
            MvcResult r = mockMvc.perform(get("/api/test-framework/runs/" + runId + "/status"))
                    .andExpect(status().isOk())
                    .andReturn();
            String status = objectMapper.readTree(r.getResponse().getContentAsString())
                    .get("status").asText();
            if ("COMPLETED".equals(status)) return;
        }
        throw new AssertionError("Run " + runId + " did not complete within 10 s");
    }

    // -------------------------------------------------------------------------
    // Tests
    // -------------------------------------------------------------------------

    @Test
    void statusEndpointReturnsOk() throws Exception {
        mockMvc.perform(get("/api/test-framework/status"))
                .andExpect(status().isOk())
                .andExpect(content().string("OK"));
    }

    @Test
    void getSuitesReturnsAllThreeSuites() throws Exception {
        mockMvc.perform(get("/api/test-framework/suites"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')]").exists())
                .andExpect(jsonPath("$[?(@.name == 'LongRunningTestSuite')]").exists())
                .andExpect(jsonPath("$[?(@.name == 'RetryTestSuite')]").exists());
    }

    @Test
    void calculatorSuiteHasExpectedTestCases() throws Exception {
        mockMvc.perform(get("/api/test-framework/suites"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')].tests[?(@.name == 'addition')]").exists())
                .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')].tests[?(@.name == 'subtraction')]").exists())
                .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')].tests[?(@.name == 'divisionByZero')]").exists());
    }

    @Test
    void retrySuiteHasExpectedTestCases() throws Exception {
        mockMvc.perform(get("/api/test-framework/suites"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[?(@.name == 'RetryTestSuite')].tests[?(@.name == 'flakyExternalCheck')]").exists())
                .andExpect(jsonPath("$[?(@.name == 'RetryTestSuite')].tests[?(@.name == 'alwaysFailingCheck')]").exists())
                .andExpect(jsonPath("$[?(@.name == 'RetryTestSuite')].tests[?(@.name == 'stableCheck')]").exists());
    }

    @Test
    void runSuiteByBodyReturnsRunId() throws Exception {
        MvcResult result = mockMvc.perform(post("/api/test-framework/suites/run/by-name")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"name\":\"LongRunningTestSuite\"}"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.runId").isNotEmpty())
                .andReturn();
        String runId = objectMapper.readTree(result.getResponse().getContentAsString())
                .get("runId").asText();
        awaitCompleted(runId);
    }

    @Test
    void runSuiteByNameReturnsRunId() throws Exception {
        MvcResult result = mockMvc.perform(post("/api/test-framework/suites/CalculatorTestSuite/run"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.runId").isNotEmpty())
                .andReturn();
        String runId = objectMapper.readTree(result.getResponse().getContentAsString())
                .get("runId").asText();
        awaitCompleted(runId);
    }

    @Test
    void runRetrySuiteReturnsRunId() throws Exception {
        MvcResult result = mockMvc.perform(post("/api/test-framework/suites/RetryTestSuite/run"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.runId").isNotEmpty())
                .andReturn();
        String runId = objectMapper.readTree(result.getResponse().getContentAsString())
                .get("runId").asText();
        awaitCompleted(runId);
    }

    @Test
    void unknownSuiteReturns404() throws Exception {
        mockMvc.perform(post("/api/test-framework/suites/NonExistentSuite/run"))
                .andExpect(status().isNotFound());
    }
}

File: kotlin-spring-boot-sample-app/src/test/kotlin/no/testframework/kotlinapp/TestFrameworkIntegrationTest.kt
package no.testframework.kotlinapp

import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestMethodOrder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc
import org.springframework.http.MediaType
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*

/**
 * Integration tests for the Kotlin Spring Boot sample app.
 *
 * Tests are ordered alphabetically so run-triggering tests complete before
 * the next test starts — avoiding 409 Conflict responses.
 */
@SpringBootTest
@AutoConfigureMockMvc
@ActiveProfiles("test")
@TestMethodOrder(MethodOrderer.MethodName::class)
class TestFrameworkIntegrationTest {

    @Autowired
    private lateinit var mockMvc: MockMvc

    private val objectMapper = ObjectMapper()

    // -------------------------------------------------------------------------
    // Helper — poll until COMPLETED (max 10 s)
    // -------------------------------------------------------------------------

    private fun awaitCompleted(runId: String) {
        repeat(100) {
            Thread.sleep(100)
            val r = mockMvc.perform(get("/api/test-framework/runs/$runId/status"))
                .andExpect(status().isOk)
                .andReturn()
            val status = objectMapper.readTree(r.response.contentAsString).get("status").asText()
            if (status == "COMPLETED") return
        }
        throw AssertionError("Run $runId did not complete within 10 s")
    }

    // -------------------------------------------------------------------------
    // Tests
    // -------------------------------------------------------------------------

    @Test
    fun statusEndpointReturnsOk() {
        mockMvc.perform(get("/api/test-framework/status"))
            .andExpect(status().isOk)
            .andExpect(content().string("OK"))
    }

    @Test
    fun getSuitesReturnsAllThreeSuites() {
        mockMvc.perform(get("/api/test-framework/suites"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')]").exists())
            .andExpect(jsonPath("$[?(@.name == 'LongRunningTestSuite')]").exists())
            .andExpect(jsonPath("$[?(@.name == 'RetryTestSuite')]").exists())
    }

    @Test
    fun calculatorSuiteHasExpectedTestCases() {
        mockMvc.perform(get("/api/test-framework/suites"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')].description")
                .value("Tests for the Calculator class"))
            .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')].tests[?(@.name == 'addition')]").exists())
            .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')].tests[?(@.name == 'subtraction')]").exists())
            .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')].tests[?(@.name == 'multiplication')]").exists())
            .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')].tests[?(@.name == 'division')]").exists())
            .andExpect(jsonPath("$[?(@.name == 'CalculatorTestSuite')].tests[?(@.name == 'divisionByZero')]").exists())
    }

    @Test
    fun retrySuiteHasExpectedTestCases() {
        mockMvc.perform(get("/api/test-framework/suites"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$[?(@.name == 'RetryTestSuite')].tests[?(@.name == 'flakyExternalCheck')]").exists())
            .andExpect(jsonPath("$[?(@.name == 'RetryTestSuite')].tests[?(@.name == 'alwaysFailingCheck')]").exists())
            .andExpect(jsonPath("$[?(@.name == 'RetryTestSuite')].tests[?(@.name == 'stableCheck')]").exists())
    }

    @Test
    fun runSuiteByBodyReturnsRunId() {
        val result = mockMvc.perform(
            post("/api/test-framework/suites/run/by-name")
                .contentType(MediaType.APPLICATION_JSON)
                .content("""{"name":"LongRunningTestSuite"}""")
        )
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.runId").isNotEmpty)
            .andReturn()
        val runId = objectMapper.readTree(result.response.contentAsString).get("runId").asText()
        awaitCompleted(runId)
    }

    @Test
    fun runSuiteByNameReturnsRunId() {
        val result = mockMvc.perform(post("/api/test-framework/suites/CalculatorTestSuite/run"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.runId").isNotEmpty)
            .andReturn()
        val runId = objectMapper.readTree(result.response.contentAsString).get("runId").asText()
        awaitCompleted(runId)
    }

    @Test
    fun runRetrySuiteReturnsRunId() {
        val result = mockMvc.perform(post("/api/test-framework/suites/RetryTestSuite/run"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.runId").isNotEmpty)
            .andReturn()
        val runId = objectMapper.readTree(result.response.contentAsString).get("runId").asText()
        awaitCompleted(runId)
    }

    @Test
    fun unknownSuiteReturns404() {
        mockMvc.perform(post("/api/test-framework/suites/NonExistentSuite/run"))
            .andExpect(status().isNotFound)
    }
}
